plugins/Unity/Launcher/xdgwatcher.cpp

```diff
@@ -0,0 +1,126 @@
+        const auto files = qdir.entryInfoList(QDir::Files);
+        for (const auto &file: files) {
+            if (file.suffix() == "desktop") {
+                m_watcher->addPath(file.absoluteFilePath());
+            }
+        }
+    }
+}
+
+// "Ubuntu style" appID is filename without versionNumber after last "_"
+const QString XdgWatcher::stripAppIdVersion(const QString rawAppID) {
+    auto appIdComponents = rawAppID.split("_");
+    appIdComponents.removeLast();
+    return appIdComponents.join("_");
+}
+
+// Standard appID see:
+// https://standards.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#desktop-file-id
+const QString XdgWatcher::toStandardAppId(const QFileInfo fileInfo) const {
+    const auto paths = QStandardPaths::standardLocations(QStandardPaths::ApplicationsLocation);
+    for (const auto &path: paths) {
+        if (fileInfo.absolutePath() == path) {
+            break;
+        }
+        if (fileInfo.absolutePath().contains(path)) {
+            auto fileStr = fileInfo.absoluteFilePath();
+            fileStr.replace(path, "");
+            fileStr.replace("/", "-");
+            fileStr.replace(".desktop", "");
+            return fileStr;
+        }
+    }
+    return fileInfo.completeBaseName();
+}
+
+const QString XdgWatcher::toAppId(const QFileInfo fileInfo) const {
+    QFile qFile(fileInfo.absoluteFilePath());
+    qFile.open(QIODevice::ReadOnly);
+    QTextStream fileStream(&qFile);
+    QString line;
+    while (fileStream.readLineInto(&line)) {
+        if (line.startsWith("X-Ubuntu-Application-ID=")) {
+            auto rawAppID = line.replace("X-Ubuntu-Application-ID=", "");
+            qFile.close();
+            return stripAppIdVersion(rawAppID);
+        }
+    }
+    qFile.close();
+
+    // If it's not an "Ubuntu" appID, we follow freedesktop standard
+    return toStandardAppId(fileInfo);
+}
+
+// Watch for newly added apps
+void XdgWatcher::onDirectoryChanged(const QString &path) {
+    const auto files = QDir(path).entryInfoList(QDir::Files);
+    const auto watchedFiles = m_watcher->files();
+    for (const auto &file: files) {
+        if (file.suffix() == "desktop" && !watchedFiles.contains(file.absoluteFilePath())) {
+            m_watcher->addPath(file.absoluteFilePath());
+
+            const auto appId = toAppId(file);
+            Q_EMIT appAdded(appId);
+        }
+    }
+}
+
+void XdgWatcher::onFileChanged(const QString &path) {
+    QFileInfo file(path);
+    if (file.exists()) {
+        // The file exists, this must be an modify event
+        const auto appId = toAppId(file);
+        Q_EMIT appInfoChanged(appId);
+    } else {
+        // File does not exist, assume this is an remove event.
+        // onDirectoryChanged will handle rename event
+
+        // As we have no way of checking the deleted file, we need to try remove both types
+        // of appIDs.
+        Q_EMIT appRemoved(toStandardAppId(file));
```

Comment:
Won't this trigger a remove event for all click packages, causing an odd race condition scenario where the entry may be removed from the app drawer, even though the app is still installed?

Reply:
no, since click will remove then create the desktop file. Since there will always be one remove and one add event. also the remove event will always pick the one first in the array/map (the oldest one).  while add will add to the bottom of the array/map.

Reply:
Hmm. OK.